src/lib/WIFI/devWIFI.cpp

```diff
@@ -796,6 +801,238 @@ static void WebUploadForceUpdateHandler(AsyncWebServerRequest *request) {
+    if (index == 0) {
+#ifdef HAS_WIFI_JOYSTICK
+        WifiJoystick::StopJoystickService();
+#endif
+        expectedFilesize = request->header("X-FileSize").toInt();
+        updatingRadio = request->header("X-Radio").toInt();
+        DBGLN("Update: '%s' size %u on radio %d", filename.c_str(), expectedFilesize, updatingRadio);
+        totalSize = 0;
+
+        // Reboot to BL mode
+        DBGLN("Reboot 1121 to bootloader mode");
+        uint8_t reboot_cmd[] = {
+            (uint8_t)(LR11XX_SYSTEM_REBOOT_OC >> 8),
+            (uint8_t)LR11XX_SYSTEM_REBOOT_OC,
+            3
+        };
+        SPIEx.write(updatingRadio, reboot_cmd, 3);
+        while(!hal.WaitOnBusy(updatingRadio)) {
+            DBGLN("Waiting...");
+            delay(10);
+        }
+
+        // Ensure we're in BL mode
+        DBGLN("Ensure BL mode");
+        packet[0] = LR11XX_BL_GET_VERSION_OC >> 8;
+        packet[1] = (uint8_t)LR11XX_BL_GET_VERSION_OC;
+        SPIEx.write(updatingRadio, packet, 2);
+        hal.WaitOnBusy(updatingRadio);
+        memset(packet, 0, sizeof(packet));
+        SPIEx.read(updatingRadio, packet, 5);
+        hal.WaitOnBusy(updatingRadio);
+        if (packet[2] != 0xDF) {
+            AsyncWebServerResponse *response = request->beginResponse(200, "application/json", R"({"status": "error", "msg": "Not in bootloader mode"})");
+            response->addHeader("Connection", "close");
+            request->send(response);
+            request->client()->close();
+            return;
+        }
+
+        // Erase flash
+        DBGLN("Erasing");
+        packet[0] = LR11XX_BL_ERASE_FLASH_OC >> 8;
+        packet[1] = (uint8_t)LR11XX_BL_ERASE_FLASH_OC;
+        SPIEx.write(updatingRadio, packet, 2);
+        while(!hal.WaitOnBusy(updatingRadio))
+        {
+            DBGLN("Waiting...");
+            delay(100);
+        }
+        DBGLN("Erased");
+
+        left_over = 0;
+        SPIEx.setHwCs(false);
+
+        pinMode(updatingRadio == SX12XX_Radio_1 ? GPIO_PIN_NSS : GPIO_PIN_NSS_2, OUTPUT);
+        digitalWrite(updatingRadio == SX12XX_Radio_1 ? GPIO_PIN_NSS : GPIO_PIN_NSS_2, HIGH);
+    }
+    if (len) {
+        DBGLN("writing %x", len);
+        // Write len bytes to LR1121 from data
+        while (len >= 256)
+        {
+            uint32_t sector_size = len > 256 ? 256 : len;
+            writeLR1121Bytes(data, sector_size);
+            totalSize += sector_size;
+            len -= sector_size - left_over;
+            data += sector_size - left_over;
+            left_over = 0;
+        }
+        memcpy(packet, data, len);
+        left_over = len;
+    }
+}
+
+static void ReadStatusForRadio(JsonObject json, SX12XX_Radio_Number_t radio)
+{
+    packet[0] = LR11XX_BL_GET_VERSION_OC >> 8;
+    packet[1] = (uint8_t)LR11XX_BL_GET_VERSION_OC;
+    SPIEx.write(radio, packet, 2);
+    hal.WaitOnBusy(radio);
+    memset(packet, 0, sizeof(packet));
+    SPIEx.read(radio, packet, 5);
+    hal.WaitOnBusy(radio);
```

Comment:
Another copy of the version getter code so it seems version getting should be its own function like
```cpp
struct PACKED lr1121VersionPacket_s
{
  uint8_t hardware;
  uint8_t type;
  uint16_t version;
};

void lr1121GetVersion(radio, lr1121VersionPacket_s *versionPacket)
{
  if (versionPacket  == nullptr)
    return;
  struct {
    uint16_t reg;
    lr1121VersionPacket_s ver;
  } buf;

  buf.reg = htobe(LR11XX_BL_GET_VERSION_OC);
  ... // do read

  *versionPacket = buf.ver;
  versionPacket->version  = be16toh(buf.ver.version);
}
```